### PR TITLE
create a macro RCL_CHECK_ALLOCATOR

### PR DESCRIPTION
**Pull request by tfoote**
to check the required methods are set on the allocator

I found 2 instances of this already partially impelemented and was starting to write several more for the time PRs. 

Linux: [Build Status](http://ci.ros2.org/job/ci_linux/3481/)
Linux arm64: [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/718/)
OSX: [Build Status](http://ci.ros2.org/job/ci_osx/2808/)
Windows: [Build Status](http://ci.ros2.org/job/ci_windows/3526/)

### CODE REVIEW

**Comment by dhood** [comment]
rcutils has a macro for this already that we could re-use: https://github.com/ros2/rcutils/blob/00e2b9db9f4b71374f1388570be2de3edabd2e37/include/rcutils/allocator.h#L111 (it looks stricter though)

**Comment by mikaelarguedas** [comment]
there is an equivalent one in rcutils that we could reuse https://github.com/ros2/rcutils/blob/00e2b9db9f4b71374f1388570be2de3edabd2e37/include/rcutils/allocator.h#L111
This basically checks that all fields are checked https://github.com/ros2/rcutils/blob/b9819514f058e32564ae6ece2da28ed64f71546d/src/allocator.c#L76-L88

**Comment by tfoote** [comment]
That's useful, shall I #define an RCL alias and use that or just call the RCUTILS one? And should we push the null argument implementation up too?

**Comment by dhood** [comment]
I'd say `#define` it rather than use the rcutils one directly, because at some point the `rcutils_allocator_t` and `rcl_allocator_t` may diverge.

Is your second question if we should de-duplicate the implementation of `RCUTILS_CHECK_ARGUMENT_FOR_NULL` and `RCL_CHECK_ARGUMENT_FOR_NULL`? I don't see a reason not to, I think the reason is just that the work pushing things to rcutils is still a bit WIP

**Comment by tfoote** [comment]
Made public and deduplicated:

Linux: [Build Status](http://ci.ros2.org/job/ci_linux/3484/)
Linux arm64: [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/719/)
OSX: [Build Status](http://ci.ros2.org/job/ci_osx/2813/)
Windows: [Build Status](http://ci.ros2.org/job/ci_windows/3527/)

**Comment by tfoote** [comment]
I reverted the merge from @mikaelarguedas as it caused instability even with the compile fix. http://ci.ros2.org/job/ci_linux/3498/ I'd suggest we merge this and then we can address the bigger question of cleaning up all usages and homogenizing all the storage of the allocator_t and using the macro everywhere in a follow up. THis will also unblock #173 to use the macros.

With the roll back it's back to the state of the last CI above.

**Comment by mikaelarguedas** [comment]
Sounds good to me, I'll work on another branch to get this consistently used by the various parts of rcl :+1:

File: rcl/src/rcl/client.c [hunk 1]
```diff
@@ -58,13 +58,7 @@ rcl_client_init(
...
   RCL_CHECK_ARGUMENT_FOR_NULL(options, RCL_RET_INVALID_ARGUMENT, rcl_get_default_allocator());
   const rcl_allocator_t * allocator = &options->allocator;
-  RCL_CHECK_FOR_NULL_WITH_MSG(
-    allocator->allocate, "allocate not set",
-    return RCL_RET_INVALID_ARGUMENT, rcl_get_default_allocator());
-  RCL_CHECK_FOR_NULL_WITH_MSG(
-    allocator->deallocate, "deallocate not set",
-    return RCL_RET_INVALID_ARGUMENT, rcl_get_default_allocator());
-
+  RCL_CHECK_ALLOCATOR(allocator, return RCL_RET_INVALID_ARGUMENT);
```
Comment (dhood) [inline rcl/src/rcl/client.c]: This won't set the error message if the allocator is invalid. That's why the changes in https://github.com/ros2/rcl/pull/178
 caused the tests to fail, because they're [expecting the error message to be set](https://github.com/ros2/rcl/blob/72bcb6b5b648b75f0c2096607c9e2fe0266352c7/rcl/test/rcl/test_node.cpp#L346). Using `RCL_CHECK_ALLOCATOR_WITH_MSG` is closer to the existing behaviour.

**Comment by tfoote** [comment]
Added msg as pointed out by @dhood and readded @mikaelarguedas 's extension.

Linux: [Build Status](http://ci.ros2.org/job/ci_linux/3520/)
Linux arm64: [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/740/) (rclpy callback count test failure believed unreleated)
OSX: [Build Status](http://ci.ros2.org/job/ci_osx/2854/)
Windows: [Build Status](http://ci.ros2.org/job/ci_windows/3577/)

File: rcl/src/rcl/publisher.c [hunk 2]
```diff
@@ -17,13 +17,13 @@ extern "C"
 {
 #endif
 
-#include "rcl/publisher.h"
```
Comment (dhood) [inline rcl/src/rcl/publisher.c]: the style of this repo has the "matching" include for the file at the top so it seems this change should be reverted

Comment (tfoote) [reply rcl/src/rcl/publisher.c]: Moved back up.

File: rcl/src/rcl/common.h [hunk 3]
```diff
@@ -23,16 +23,6 @@ extern "C"
 #include "rcl/error_handling.h"
 #include "rcl/types.h"
 
-#define RCL_CHECK_ARGUMENT_FOR_NULL(argument, error_return_type, allocator) \
```
Comment (dhood) [inline rcl/src/rcl/common.h]: For my understanding, is it intentional that the files that use `RCL_CHECK_ARGUMENT_FOR_NULL` can only do so because it's implicitly included in `common.h`? FWIU we have an "include what you use" policy that might suggest those files (e.g. `client.c`) should now include `rcl/error_handling.h` explicitly, but the purpose of `common.h` might be to get around that rule in some circumstances.

Comment (tfoote) [reply rcl/src/rcl/common.h]: I don't think that was the point of common.h I removed the include from common.h and filled in every source file that had an issue, then added it back so common.c could compile and link correctly.

Comment (dhood) [reply rcl/src/rcl/common.h]: it should be fine to remove it from common.h now right? or could you give more info on why it's still needed?

Comment (mikaelarguedas) [reply rcl/src/rcl/common.h]: I believe it needs the types as it returns some `rcl_ret_t`

Comment (dhood) [reply rcl/src/rcl/common.h]: they should come from rcl/types.h no?

Comment (mikaelarguedas) [reply rcl/src/rcl/common.h]: yes exactly, all I meant is currently common.c doesnt include types.h

Comment (mikaelarguedas) [reply rcl/src/rcl/common.h]: Looks like I misunderstood what we're talking about here. If now all files include `rcl/error_handling.h` (including `common.c`) there should indeed be no need to include it in `common.h`. Not sure what the linking issue you faced was from

Comment (tfoote) [reply rcl/src/rcl/common.h]: The linking issue was due to one more missing include. I had to trace back to an `implicit declaration of function` that was caused by the missing macro. I cleared the include from `common.h` in 
a5e5825

**Comment by tfoote** [comment]
Thanks for the reviews. Here's a sanity check that the header cleanup was fine. 

Linux: [Build Status](http://ci.ros2.org/job/ci_linux/3524/)

**Comment by dhood** [comment]
Looks like there are some files (not all) that don't need to include common.h anymore because they only needed the argument checking macros: I added https://github.com/ros2/rcl/pull/176/commits/3fef7c147bcfb1ccd9f33d050549ddbc349ebf5b to this PR

**Comment by tfoote** [comment]
Thanks one more CI on top of that: 

Linux: [Build Status](http://ci.ros2.org/job/ci_linux/3528/)
